### PR TITLE
OCPCLOUD-2431: rebase on upstream 1.29.0 release

### PR DESCRIPTION
**Pull request by elmiko**
This commit rebases the autoscaler on top of the Kubernetes/Autoscaler 1.29.0 release. There are several commits that we carry on top of the upstream autoscaler and the rebase process allows us to preserve those. Here is a description of the process I used to create this PR.

(inspired by the commit description for the 1.18 rebase. [pr #139](https://github.com/openshift/kubernetes-autoscaler/pull/139))

**Process**

First we need to identify the carry commits that we currently have, this is done against our previous rebase to catch new changes. Once identified we will drop commits which have merged upstream and only carry unique commits. (see below for the carried and dropped commits).

Identify carry commits (run from the openshift/master branch), these are the commits that begin with `UPSTREAM: ` up until the merge commit for the previous rebase commit (`merge upstream/cluster-autoscaler-1.28.0`)
```
git log -n 30 --oneline --no-merges
```

After identifying the carry commits, the next step is to create the new commit-tree that will be used for the rebase and then cherry pick the carry commits into the new branch. The following commands cover these steps:
```
$ git remote update # make sure we update our refs
$ git checkout cluster-autoscaler-1.29.0
$ git checkout -b merge-tmp # create a temporary branch for our merge commit
$ git checkout openshift/master # we want to be at the tip of the openshift master branch when we run the next command
$ echo 'merge upstream/cluster-autoscaler-1.29.0' | git commit-tree merge-tmp^{tree} -p HEAD -p merge-tmp -F - # create a new merge commit for our history
deadbeef12345678 # id of new merge commit
$ git branch merge-1.29 deadbeef12345678 # create a new branch for the cherry-pick work
$ git checkout merge-1.29
$ git cherry-pick <carry commits> # cherry pick the needed commits into the new branch
```

With the `merge-1.29` branch in place, I cherry picked the carry commits which applied, resolved merge conflicts, and finally tested the resulting tree against the unit test and end-to-end suite.

**Carried Commits**

These commits are for features which have not yet been accepted upstream, are integral to our CI platform, or are specific to the releases we create for OpenShift.
```
2120c26a5 UPSTREAM: <drop>: Temporarily reintroduce the flag --scale-up-from-zero-default-arch
10ff5b566 UPSTREAM: 6491: add informer argument to clusterapi provider builder
efeb8e308 UPSTREAM: <carry>: Fix unstructured taint parsing in Cluster API provider
1f937037f UPSTREAM: <carry>: revert capacity annotations
b3faab26a UPSTREAM: <carry>: add machine api label and taint functionality
5ad4b142e UPSTREAM: <carry>: update os::util::list_test_packages_under
cca0d470b UPSTREAM: <carry>: remove deprecated go test command
9ce518c79 UPSTREAM: <carry>: Have VPA ignore phantom containers named "POD"
9b1d287f3 UPSTREAM: <carry>: Handle old Machine API specific machine delete annotation
c3ff7aaad UPSTREAM: <carry>: Rename FailureMessage to ErrorMessage
fcc7648db UPSTREAM: <carry>: configure repository for OpenShift releases
```

**Dropped Commits**

These commits were squashed into the `fcc7648db` commit.
```
d017ad320 UPSTREAM: <carry>: Updating atomic-openshift-cluster-autoscaler-container image to be consistent with ART for 4.16 Reconciling with https://github.com/openshift/ocp-build-data/tree/79948f91df396fed6784d24a983f2a6acb2c4fb8/images/atomic-openshift-cluster-autoscaler.yml
e239872a0 UPSTREAM: <carry>: Updating ose-vertical-pod-autoscaler-container image to be consistent with ART for 4.16 Reconciling with https://github.com/openshift/ocp-build-data/tree/17f81aa502d841340a0b97ae1365568f8253b6ec/images/vertical-pod-autoscaler.yml
dd8ccf7d9 UPSTREAM: <carry>: Updating atomic-openshift-cluster-autoscaler-container image to be consistent with ART for 4.16 Reconciling with https://github.com/openshift/ocp-build-data/tree/b7ba6c3881993b1c2e0a96905cce738d604e1f1d/images/atomic-openshift-cluster-autoscaler.yml
908ccaaf9 UPSTREAM: <carry>: Add Snyk file to exclude vendor directory on scan
57d3f0f3e UPSTREAM: <carry>: Updating ose-vertical-pod-autoscaler-container image to be consistent with ART Reconciling with https://github.com/openshift/ocp-build-data/tree/4022cd290f00a44d667dda03f2d78d84a488c7ed/images/vertical-pod-autoscaler.yml
050999591 UPSTREAM: <carry>: Sync OWNERS file
67cfac729 UPSTREAM: <carry>: Updating ose-vertical-pod-autoscaler-container image to be consistent with ART Reconciling with https://github.com/openshift/ocp-build-data/tree/e5df10e41b14c724468c8f78e2690f219279bfc4/images/vertical-pod-autoscaler.yml
1f412862e UPSTREAM: <carry>: Updating ose-vertical-pod-autoscaler-container image to be consistent with ART Reconciling with https://github.com/openshift/ocp-build-data/tree/79604dfeca884ebe3d33791aaca81b0a77d21d2c/images/vertical-pod-autoscaler.yml
db61a2ee9 UPSTREAM: <carry>: Add vendor for balancer
```

This commit was dropped
```
efea62a0b UPSTREAM: 6066: Allow overriding the kubernetes.io/arch label set by the scale from zero methods via a new cmdline arg
```

Of special note in this rebase is this commit
```
c74af5632 UPSTREAM: <carry>: revert capacity annotations
```
due to the scale from zero changes being accepted upstream we can now drop our carried patch. but, the upstream implementation has differed slightly from our's (mainly around annotation names). we will need to carry this patch until we can fix all the providers to properly use the new annotations. This patch should be able to be dropped in the 4.17 release of openshift, see https://issues.redhat.com/browse/OCPCLOUD-944.

### CODE REVIEW

**Comment by elmiko** [comment]
looks like we will need to update the go version to make this work

**Comment by elmiko** [comment]
looks like still some issues around building the image and go version

**Comment by JoelSpeed** [comment]
Compilation failures will need looking into

**Comment by aleskandro** [comment]
Hi @elmiko @JoelSpeed as we were discussing in openshift/cluster-autoscaler-operator#311, I prepared this commit to overcome the failure that should arise when the operator tries to deploy the autoscaler with the `--scale-up-from-zero-default-arch` flag, dropped in this rebase PR.

We could cherry-pick https://github.com/aleskandro/kubernetes-autoscaler/commit/b435ec8fe8eed9886b10e99100af018a3694f247 and let it be dropped as a follow-up PR (or at most by the next rebase). After this cherry-pick and the rebase, we can safely merge openshift/cluster-autoscaler-operator#311.

**Comment by elmiko** [comment]
@aleskandro i was looking at your pr as i was cherry-picking the commits, it seems like the code to support the environment variable is in here, do we need an extra patch?

**Comment by elmiko** [comment]
i read the CAO thread, i think we should just add the no-op commit as a drop here.

**Comment by aleskandro** [comment]
Hey @elmiko, the operator is still setting the --scale-up-from-zero-default-arch parameter in the Args of the  autoscaler's deployment, but with the rebase, this parameter is not defined anymore. I'm getting a doubt now as I don't remember anymore the tests that I did at the time: can the autoscaler accept unrecognised command line arguments?

**Comment by elmiko** [comment]
adding a fix for the goimports failure

**Comment by elmiko** [comment]
looks like the failures are coming from the missing flag now, i will add that patch tomorrow.

**Comment by elmiko** [comment]
i added @aleskandro 's no-op patch as a drop

**Comment by elmiko** [comment]
tests looking good

File: .snyk [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# References:
+# https://docs.snyk.io/scan-applications/snyk-code/using-snyk-code-from-the-cli/excluding-directories-and-files-from-the-snyk-code-cli-test
+# https://docs.snyk.io/snyk-cli/commands/ignore
+exclude:
+  global:
+    - vendor/**
```
Comment (JoelSpeed) [inline .snyk]: This needs quoting, can we fix it up?

**Comment by JoelSpeed** [comment]
Fixup is minor
Lets merge and sort later
/lgtm
/approve

**Pull request merged**